ONSdigital/eq-author-app
# EAR 1065 - Improve validation performance (large surveys)

### What is the context of this PR?

Currently Author API responds very slowly to graphql requests for large questionnaires - cloning the Low Carbon Survey locally (and using it on prod) allows observing response times as great as three seconds per query.

This PR improves Author-API's response times significantly for large surveys such as LCS by optimising the questionnaire validation code.

For more details see my report here: [Profiling & optimising validation performance](https://tbaggaley.github.io/ajv_perf).

NB: I ran the benchmarks and profiling on Martyn's new MongoDB branch, hence the lower overall times in the benchmarks in the report compared to current `master`; I think the mongo branch is better performing overall.

### How to review

- First open LCS on prod or import it locally using the `master` branch and observe the high response times.
- Switch to this branch and import LCS.
- Refresh the page, switch pages, etc. to generate network activity and observe the graphql response times.
- Should be significantly less (generally 50% less or more).

### What to do after everything is green

1. - [ ] Bring this branch up-to-date with Origin/Master
2. - [ ] If there are a lot of commits or some are not helpful to read, squash them down
3. - [ ] Click the **Merge** button on this pull request
4. - [ ] Does this change mean the **Capability examples** questionnaire should be updated?
5. - [ ] Move the Jira ticket for this task into the next stage of the process
